File: Yarn/measurement.go
package yarn

import (
	"time"

	"github.com/google/uuid"
)

// Measurement contains conveyance metrics from a single agent interaction.
type Measurement struct {
	ID        string    `json:"id"`
	Timestamp time.Time `json:"timestamp"`

	// Session context
	SessionID      string `json:"session_id"`
	ConversationID string `json:"conversation_id"`
	TurnNumber     int    `json:"turn_number"`

	// Participants
	SenderID   string `json:"sender_id"`
	SenderName string `json:"sender_name"`
	SenderRole string `json:"sender_role"`

	ReceiverID   string `json:"receiver_id"`
	ReceiverName string `json:"receiver_name"`
	ReceiverRole string `json:"receiver_role"`

	// Boundary objects (hidden states)
	SenderHidden   *HiddenState `json:"sender_hidden,omitempty"`
	ReceiverHidden *HiddenState `json:"receiver_hidden,omitempty"`

	// Core conveyance metrics
	DEff      int     `json:"d_eff"`      // Effective dimensionality
	Beta      float64 `json:"beta"`       // Collapse indicator
	Alignment float64 `json:"alignment"`  // Cosine similarity
	CPair     float64 `json:"c_pair"`     // Bilateral conveyance

	// Quality indicators
	BetaStatus   BetaStatus `json:"beta_status"`
	IsUnilateral bool       `json:"is_unilateral"`

	// Message context
	MessageContent string `json:"message_content,omitempty"`
	TokenCount     int    `json:"token_count,omitempty"`
}

// BetaStatus indicates the quality of the β value.
// Beta (β) is the collapse indicator from the Conveyance Framework.
// Lower values indicate better dimensional preservation.
type BetaStatus string

const (
	BetaOptimal    BetaStatus = "optimal"    // β ∈ [1.5, 2.0) - ideal range
	BetaMonitor    BetaStatus = "monitor"    // β ∈ [2.0, 2.5) - acceptable, watch for drift
	BetaConcerning BetaStatus = "concerning" // β ∈ [2.5, 3.0) - dimensional compression detected
	BetaCritical   BetaStatus = "critical"   // β ≥ 3.0 - severe collapse, intervention needed
	BetaUnknown    BetaStatus = "unknown"    // β ≤ 0 or β ∈ (0, 1.5) - invalid or uncategorized
)

// IsValid returns true if this is a valid beta status.
func (s BetaStatus) IsValid() bool {
	switch s {
	case BetaOptimal, BetaMonitor, BetaConcerning, BetaCritical, BetaUnknown:
		return true
	default:
		return false
	}
}

// NewMeasurement creates a new Measurement with a generated ID.
// Note: A SenderID or ReceiverID must be set before the measurement is valid.
// Use SetSender() or SetReceiver() to configure participants after creation.
func NewMeasurement() *Measurement {
	return &Measurement{
		ID:        uuid.New().String(),
		Timestamp: time.Now(),
	}
}

// NewMeasurementForTurn creates a measurement for a specific conversation turn.
func NewMeasurementForTurn(sessionID, convID string, turn int) *Measurement {
	m := NewMeasurement()
	m.SessionID = sessionID
	m.ConversationID = convID
	m.TurnNumber = turn
	return m
}

// SetSender sets the sender information.
func (m *Measurement) SetSender(id, name, role string, hidden *HiddenState) {
	m.SenderID = id
	m.SenderName = name
	m.SenderRole = role
	m.SenderHidden = hidden
}

// SetReceiver sets the receiver information.
func (m *Measurement) SetReceiver(id, name, role string, hidden *HiddenState) {
	m.ReceiverID = id
	m.ReceiverName = name
	m.ReceiverRole = role
	m.ReceiverHidden = hidden
}

// IsBilateral returns true if both sender and receiver have hidden states.
func (m *Measurement) IsBilateral() bool {
	hasSender := m.SenderHidden != nil && len(m.SenderHidden.Vector) > 0
	hasReceiver := m.ReceiverHidden != nil && len(m.ReceiverHidden.Vector) > 0
	return hasSender && hasReceiver
}

// Validate checks if the measurement is valid.
// Returns a ValidationError if invalid, nil if valid.
func (m *Measurement) Validate() *ValidationError {
	// Required fields
	if m.ID == "" {
		return &ValidationError{Field: "id", Message: "id is required"}
	}
	if m.Timestamp.IsZero() {
		return &ValidationError{Field: "timestamp", Message: "timestamp is required"}
	}

	// TurnNumber must be non-negative
	if m.TurnNumber < 0 {
		return &ValidationError{Field: "turn_number", Message: "turn_number must be non-negative"}
	}

	// At least one participant is required
	if m.SenderID == "" && m.ReceiverID == "" {
		return &ValidationError{Field: "sender_id", Message: "at least one of sender_id or receiver_id is required"}
	}

	// Metric validations
	if m.DEff < 0 {
		return &ValidationError{Field: "d_eff", Message: "d_eff must be non-negative"}
	}
	if m.Beta < 0 {
		return &ValidationError{Field: "beta", Message: "beta must be non-negative"}
	}
	if m.Alignment < -1 || m.Alignment > 1 {
		return &ValidationError{Field: "alignment", Message: "alignment must be in range [-1, 1]"}
	}

	// BetaStatus validation (empty is allowed as default)
	if m.BetaStatus != "" && !m.BetaStatus.IsValid() {
		return &ValidationError{Field: "beta_status", Message: "invalid beta_status"}
	}

	// Validate hidden states if present
	if err := m.SenderHidden.Validate(); err != nil {
		return &ValidationError{Field: "sender_hidden." + err.Field, Message: err.Message}
	}
	if err := m.ReceiverHidden.Validate(); err != nil {
		return &ValidationError{Field: "receiver_hidden." + err.Field, Message: err.Message}
	}

	return nil
}

// ComputeBetaStatus determines the status based on β value.
func ComputeBetaStatus(beta float64) BetaStatus {
	switch {
	case beta <= 0:
		return BetaUnknown
	case beta < 1.5:
		return BetaUnknown // Explicitly handle (0, 1.5) range
	case beta < 2.0:
		return BetaOptimal // β ∈ [1.5, 2.0)
	case beta < 2.5:
		return BetaMonitor // β ∈ [2.0, 2.5)
	case beta < 3.0:
		return BetaConcerning // β ∈ [2.5, 3.0)
	default:
		return BetaCritical // β ≥ 3.0
	}
}

File: web-ui/src/types/measurement.ts
/**
 * Measurement types matching Yarn/measurement.go
 * and Weaver/pkg/api/events.go
 *
 * Note: Field names use camelCase to match the API JSON responses.
 */

/**
 * BetaStatus indicates the quality of the beta value.
 * Beta is the collapse indicator from the Conveyance Framework.
 * Lower values indicate better dimensional preservation.
 */
export type BetaStatus = 'optimal' | 'monitor' | 'concerning' | 'critical' | 'unknown';

/**
 * Beta status thresholds and ranges.
 * - optimal: beta in [1.5, 2.0) - ideal range
 * - monitor: beta in [2.0, 2.5) - acceptable, watch for drift
 * - concerning: beta in [2.5, 3.0) - dimensional compression detected
 * - critical: beta >= 3.0 - severe collapse, intervention needed
 * - unknown: beta <= 0 or in (0, 1.5) - invalid or uncategorized
 */
export const BETA_STATUS_RANGES = {
  optimal: { min: 1.5, max: 2.0 },
  monitor: { min: 2.0, max: 2.5 },
  concerning: { min: 2.5, max: 3.0 },
  critical: { min: 3.0, max: Infinity },
} as const;

/**
 * Computes the BetaStatus based on beta value.
 * Mirrors the Go ComputeBetaStatus function.
 */
export function computeBetaStatus(beta: number): BetaStatus {
  if (beta <= 0) return 'unknown';
  if (beta < 1.5) return 'unknown';
  if (beta < 2.0) return 'optimal';
  if (beta < 2.5) return 'monitor';
  if (beta < 3.0) return 'concerning';
  return 'critical';
}

/** Check if a BetaStatus is valid. */
export function isValidBetaStatus(status: string): status is BetaStatus {
  return ['optimal', 'monitor', 'concerning', 'critical', 'unknown'].includes(status);
}

/**
 * HiddenState represents the boundary object - semantic state before text projection.
 * Maps to HiddenState in Yarn/message.go.
 *
 * Memory note: Vector can be large (e.g., 4096 floats = 16KB for typical LLMs).
 * For models with larger hidden dimensions (e.g., 8192), expect ~32KB per state.
 */
export interface HiddenState {
  /** Hidden state vector, typically 2048-8192 float32 values */
  vector: number[];
  /** Original tensor shape, e.g., [1, seq_len, hidden_dim] */
  shape: number[];
  /** Layer index this state was extracted from */
  layer: number;
  /** Data type, typically "float32" */
  dtype: string;
}

/** Get the hidden dimension size from a HiddenState. */
export function getHiddenDimension(state: HiddenState | null | undefined): number {
  if (!state) return 0;
  if (state.shape.length < 2) return state.vector.length;
  return state.shape[state.shape.length - 1];
}

/**
 * Measurement contains conveyance metrics from a single agent interaction.
 * Maps to Measurement in Yarn/measurement.go.
 */
export interface Measurement {
  id: string;
  timestamp: string;

  // Session context
  sessionId: string;
  conversationId: string;
  turnNumber: number;

  // Participants
  senderId: string;
  senderName: string;
  senderRole: string;

  receiverId: string;
  receiverName: string;
  receiverRole: string;

  // Boundary objects (hidden states)
  senderHidden?: HiddenState | null;
  receiverHidden?: HiddenState | null;

  // Core conveyance metrics
  /** Effective dimensionality */
  dEff: number;
  /** Collapse indicator */
  beta: number;
  /** Cosine similarity */
  alignment: number;
  /** Bilateral conveyance */
  cPair: number;

  // Quality indicators
  betaStatus: BetaStatus;
  isUnilateral: boolean;

  // Message context
  messageContent?: string;
  tokenCount?: number;
}

/** Check if a measurement is bilateral (both sender and receiver have hidden states). */
export function isBilateral(measurement: Measurement): boolean {
  const hasSender = measurement.senderHidden && measurement.senderHidden.vector.length > 0;
  const hasReceiver = measurement.receiverHidden && measurement.receiverHidden.vector.length > 0;
  return Boolean(hasSender && hasReceiver);
}

/**
 * MeasurementEvent provides a richer measurement event structure.
 * Maps to MeasurementEvent in events.go.
 */
export interface MeasurementEvent {
  // Core identification
  id: string;
  timestamp: string;

  // Session context
  sessionId?: string;
  conversationId?: string;
  turn: number;

  // Participants
  senderId?: string;
  senderName?: string;
  senderRole?: string;

  receiverId?: string;
  receiverName?: string;
  receiverRole?: string;

  // Core conveyance metrics
  /** Effective dimensionality */
  deff: number;
  /** Collapse indicator */
  beta: number;
  /** Cosine similarity */
  alignment: number;
  /** Bilateral conveyance */
  cpair: number;

  // Quality indicators
  betaStatus?: BetaStatus;
  isUnilateral?: boolean;

  // Message context
  tokenCount?: number;
}

/**
 * MeasurementBatchEvent contains multiple measurements for batch updates.
 * Maps to MeasurementBatchEvent in events.go.
 */
export interface MeasurementBatchEvent {
  measurements: MeasurementEvent[];
  sessionId?: string;
  turnRange: [number, number]; // [start, end]
}

/**
 * BetaAlertEvent is sent when beta reaches concerning or critical levels.
 * Maps to BetaAlertEvent in events.go.
 */
export interface BetaAlertEvent {
  measurementId: string;
  sessionId?: string;
  turn: number;
  beta: number;
  status: BetaStatus;
  previousBeta?: number;
  previousStatus?: BetaStatus;
  alertMessage: string;
  timestamp: string;
}

/**
 * ConversationTurnEvent is sent for each conversation turn with context.
 * Maps to ConversationTurnEvent in events.go.
 */
export interface ConversationTurnEvent {
  sessionId: string;
  conversationId: string;
  turn: number;
  senderName: string;
  receiverName: string;
  timestamp: string;
}

/**
 * MeasurementData is a simplified measurement structure for charts.
 * Used for backward compatibility with simple data displays.
 */
export interface MeasurementData {
  turn: number;
  deff: number;
  beta: number;
  alignment: number;
  cpair: number;
  sender?: string;
  receiver?: string;
}
